lib/client.js

```diff
@@ -1226,21 +1110,39 @@ async function connect (client) {
 
       const session = http2.connect(client[kUrl], {
         createConnection: () => socket,
-        peerMaxConcurrentStreams: client[kHTTP2SessionState].maxConcurrentStreams
+        peerMaxConcurrentStreams: client[kMaxConcurrentStreams]
       })
 
       client[kHTTPConnVersion] = 'h2'
-      session[kClient] = client
-      session[kSocket] = socket
-      session.on('error', onHttp2SessionError)
-      session.on('frameError', onHttp2FrameError)
-      session.on('end', onHttp2SessionEnd)
-      session.on('goaway', onHTTP2GoAway)
+      session.on('error', function (err) {
+        assert(err.code !== 'ERR_TLS_CERT_ALTNAME_INVALID')
+
+        this[kError] = err
+
+        onError(this[kClient], err)
+      })
+      session.on('frameError', function (type, code, id) {
+        if (id === 0) {
+          const err = new InformationalError(`HTTP/2: "frameError" received - type ${type}, code ${code}`)
+          util.destroy(this, err)
+          util.destroy(socket, err)
+        }
+      })
+      session.on('end', function () {
+        const err = new SocketError('other side closed')
+        util.destroy(this, err)
+        util.destroy(socket, err)
+      })
+      session.on('goaway', function (code) {
+        this[kError] = new InformationalError(`HTTP/2: "GOAWAY" frame received with code ${code}`)
+        onSocketClose.call(session)
+      })
       session.on('close', onSocketClose)
       session.unref()
 
-      client[kHTTP2Session] = session
-      socket[kHTTP2Session] = session
+      session[kOpenStreams] = 0
+
+      client[kSocket] = session
```

Comment:
Do you think it can have potential side-effects?
They are a duplex stream, so they share APIs but the way that they are used is different 🤔

Reply:
I *think* in the way it is currently used it does not matter.